src/platform/datapath_winrdma.c

```diff
@@ -0,0 +1,5151 @@
+
+    Status = RdmaConnection->QueuePair->lpVtbl->SendWithImmediate(
+        RdmaConnection->QueuePair,
+        RequestContext,
+        Sge,
+        SgeSize,
+        Flags,
+        ImmediateData);
+    if (QUIC_FAILED(Status))
+    {
+        QuicTraceLogError(
+            NDSPIReadFailed,
+            "[ ndspi] NdspiSendWithImmediate failed, status: %d", Status);
+    }
+
+    return Status;
+}
+
+//
+// Post RDMA Receive
+//
+_IRQL_requires_max_(PASSIVE_LEVEL)
+QUIC_STATUS
+NdspiPostReceive(
+    _Inout_ RDMA_CONNECTION* RdmaConnection,
+    _In_opt_ VOID* Context,
+    __in_ecount_opt(SgeSize) const ND2_SGE *Sge,
+    _In_ ULONG SgeSize
+    )
+{
+    QUIC_STATUS Status = QUIC_STATUS_SUCCESS;
+
+    if (!RdmaConnection ||
+        !RdmaConnection->QueuePair)
+    {
+        Status = ND_INVALID_PARAMETER;
+        QuicTraceLogError(
+            NdspiPostReceiveFailed,
+            "[ ndspi] NdspiSendWithImmediate failed, status: %d", Status);
+        return QUIC_STATUS_INVALID_PARAMETER;
+    }
+
+    Status = RdmaConnection->QueuePair->lpVtbl->Receive(
+        RdmaConnection->QueuePair,
+        Context,
+        Sge,
+        SgeSize);
+    if (QUIC_FAILED(Status))
+    {
+        QuicTraceLogError(
+            NDSPIReadFailed,
+            "[ ndspi] NdspiPostReceive failed, status: %d", Status);
+    }
+
+    return Status;
+}
+
+//
+// Creates an RDMA  Initialization context
+//
+_IRQL_requires_max_(PASSIVE_LEVEL)
+QUIC_STATUS
+CxPlatRdmaAdapterInitialize(
+    _In_ const QUIC_ADDR*           LocalAddress,
+    _Out_ void**                    Adapter           
+    )
+
+{
+    *Adapter = NULL;
+
+    SOCKADDR sockAddr = {0};
+    QUIC_STATUS Status = QUIC_STATUS_SUCCESS;
+
+    RDMA_NDSPI_ADAPTER* RdmaAdapter = NULL;
+
+    if (LocalAddress->si_family == AF_INET) {
+        // IPv4 address
+        const SOCKADDR_IN* sockAddrIn = &LocalAddress->Ipv4;
+        memcpy(&sockAddr, sockAddrIn, sizeof(SOCKADDR_IN));
+    }
+    else
+    {
+        // IPv6 address
+        const SOCKADDR_IN6* sockAddrIn6 = &LocalAddress->Ipv6;
+        memcpy(&sockAddr, sockAddrIn6, sizeof(SOCKADDR_IN6));
+    }
+
+    RdmaAdapter = (RDMA_NDSPI_ADAPTER*)CXPLAT_ALLOC_PAGED(sizeof(RDMA_NDSPI_ADAPTER), QUIC_POOL_DATAPATH);
+    if (RdmaAdapter == NULL)
+    {
+        QuicTraceEvent(
+            AllocFailure,
+            "Allocation of '%s' failed. (%llu bytes)",
+            "RDMA_NDSPI_ADAPTER",
+            sizeof(RDMA_NDSPI_ADAPTER));
+        return QUIC_STATUS_OUT_OF_MEMORY;
+    }
+
+    Status = NdOpenAdapter(
+        &IID_IND2Adapter,
+        &sockAddr,
+        sizeof(sockAddr),
+        &RdmaAdapter->Adapter);
+
+    if (QUIC_FAILED(Status)) {
+        QuicTraceLogError(
+            NdOpenAdapterFailed,
+            "NdOpenAdapter failed, status: %d", Status);
+        goto ErrorExit;
+    }
+
+    Status = NdspiCreateOverlappedFile(RdmaAdapter, &RdmaAdapter->OverlappedFile);
+    if (QUIC_FAILED(Status)) {
+        QuicTraceLogError(
```

Comment:
Here (and in other occurences with wrappers), we end up logging the same error multiple time. It might be worth chosing between loging in the wrapper or out of it and systematically apply it.